File: 카카오/파괴되지_않은_건물.cpp
// https://school.programmers.co.kr/learn/courses/30/lessons/92344
#include <string>
#include <vector>

using namespace std;
int sum[1002][1002];

int solution(vector<vector<int>> board, vector<vector<int>> skills) {
    int n = board.size(), m = board[0].size(), result = 0;
    for (auto skill : skills) {
        int change = skill[0] == 1 ? -skill[5] : skill[5];
        sum[skill[1]][skill[2]] += change;
        sum[skill[3] + 1][skill[2]] -= change;
        sum[skill[1]][skill[4] + 1] -= change;
        sum[skill[3] + 1][skill[4] + 1] += change;
    }
    
    for (int i = 1; i < n; i++) 
        for (int j = 0; j < m; j++) 
            sum[i][j] += sum[i - 1][j];
    
    for (int j = 1; j < m; j++) 
        for (int i = 0; i < n; i++) 
            sum[i][j] += sum[i][j - 1];
    
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < m; j++) {
            if (board[i][j] + sum[i][j] > 0)
                result++;
        }
    }
    
    return result;
}

File: 카카오/파괴되지_않은_건물.java
import java.util.*;

class Solution {
    public int solution(int[][] board, int[][] skills) {
        int n = board.length;
        int m = board[0].length;
        int[][] changes = new int[n][m];
        for (int[] skill : skills) {
            int r1 = skill[1];
            int c1 = skill[2];
            int r2 = skill[3];
            int c2 = skill[4];
            int degree = skill[5];
            if (skill[0] == 1) {
                degree *= -1;
            }
            changes[r1][c1] += degree;
            if (r2 + 1 < n) {
                changes[r2 + 1][c1] -= degree;
                if (c2 + 1 < m)
                    changes[r2 + 1][c2 + 1] += degree;
            }
            if (c2 + 1 < m) {
                changes[r1][c2 + 1] -= degree;
            }
        } 
        int count = 0;
        for (int i = 0; i < n; i++) {
            for (int j = 1; j < m; j++) {
                changes[i][j] += changes[i][j - 1];
            }
        }
        for (int j = 0; j < m; j++) {
            for (int i = 1; i < n; i++) {
                changes[i][j] += changes[i - 1][j];
            }
        }
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (changes[i][j] + board[i][j] > 0)
                    count++;
            }
        }
        return count;
    }
}